File: swift/LeetCodeProblems/LeetCodeProblems/DataStructures/ListNode.swift
//
//  ListNode.swift
//  LeetCodeProblems
//
//  Created by Aynur Nasybullin on 08.09.2023.
//

import Foundation

public final class ListNode {
    public var val: Int
    public var next: ListNode?
    
    public init() {
        self.val = 0
        self.next = nil
    }
    
    public init(_ val: Int) {
        self.val = val
        self.next = nil
    }
    
    public init(_ val: Int, _ next: ListNode?) {
        self.val = val
        self.next = next
    }
}

extension ListNode : CustomDebugStringConvertible {
    public var debugDescription: String {
        var tmp: ListNode? = self
        var description = "["
        
        while tmp != nil {
            if let val = tmp?.val {
                description += "\(String(describing: val)), "
            }
            tmp = tmp?.next
        }
        
        description.removeLast()
        description += "]"
        
        return description
    }
}

File: cpp/data_structures/list_node.cpp
#include "list_node.h"

ListNode::ListNode() : val(0), next(nullptr) { };

ListNode::ListNode(int x) : val(x), next(nullptr) { };

ListNode::ListNode(int x, ListNode *next) : val(x), next(next) { };

bool ListNode::Equals(const ListNode *l1, const ListNode *l2) {
    while (l1 != nullptr && l2 != nullptr) {
        if (l1->val != l2->val) { return false; }

        l1 = l1->next;
        l2 = l2->next;
    }

    return l1 == nullptr && l2 == nullptr;
};